cat/spirv.cat

```diff
@@ -0,0 +1,127 @@
+
+//let chains = (M | F | CBAR) * (M | F | CBAR)
+//let chains = [M | F | CBAR]
+
+let avsg = [AV & SG]
+let avwg = ((avsg; (hb & ssg & avvisinc))?); [AV & WG]
+let avqf = ((avsg; (hb & ssg & avvisinc))? ; (avwg; (hb & swg & avvisinc))?); [AV & QF]
+let avsh = ((avsg; (hb & ssg & avvisinc))? ; (avwg; (hb & swg & avvisinc))?; (avqf; (hb & sqf & avvisinc))?); [AV & SHADER]
+let avdv = [AV & DV]
+
+let vissg = [VIS & SG]
+let viswg = [VIS & WG]; (((hb & ssg & avvisinc); vissg)?)
+let visqf = [VIS & QF]; (((hb & ssg & avvisinc) ; viswg)?; ((hb & ssg & avvisinc); vissg)?)
+let vissh = [VIS & SHADER]; (((hb & ssg & avvisinc) ; visqf)?; ((hb & swg & avvisinc); viswg)?; ((hb & sqf & avvisinc); vissg)?)
+let visdv = [VIS & DV]
+
+// loc order, relates memory accesses to the same location
+let locord = loc &
+             (  (hb & int & vloc) |
+                ([R \ PRIV]; hb; [R | W \ PRIV]) | // RaR, WaR (non-private)
+                ([R]; ssw+; [R | W]) | // RaR, WaR (any)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsg; (hb & ssg);                         [W \ PRIV])) |                                 // WaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsg; (hb & ssg); vissg; (po?; avvisinc); [R \ PRIV]; (po? & avvisinc); [R \ PRIV])) |   // RaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avwg; (hb & swg);                         [W \ PRIV])) |                                 // WaW (via subgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avwg; (hb & swg); viswg; (po?; avvisinc); [R \ PRIV]; (po? & avvisinc); [R \ PRIV])) |   // RaW (via workgroup instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avqf; (hb & sqf);                         [W \ PRIV])) |                                 // WaW (via queue family instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avqf; (hb & sqf); visqf; (po?; avvisinc); [R \ PRIV]; (po? & avvisinc); [R \ PRIV])) |   // RaW (via queue family instance domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsh; (hb);                               [W \ PRIV])) |                                 // WaW (via shader domain)
+                (vloc & ([W \ PRIV]; (po? & avvisinc); avsh; (hb);       vissh; (po?; avvisinc); [R \ PRIV]; (po? & avvisinc); [R \ PRIV])) |   // RaW (via shader domain)
+                (       ([W];        (hb & avvisinc);  avdv; (hb);                               [W])) |                                        // WaW (via device domain)
+                (       ([W];        (hb & avvisinc);  avdv; (hb);       visdv; (hb & avvisinc); [R]; (po? & avvisinc); [R \ PRIV]))            // RaW (via device domain)
+             )
+
+let w-locord = [W]; locord
+
+let visto = (w-locord \ (w-locord; w-locord)); [R] // visible to = location ordered W->R with no intervening write (W->W->R)
```

Comment:
`imm[r] = r \ (r;r+)` ... you are missing the `+`

Reply:
Added